functional_programming/06_applicatives.py

```diff
@@ -0,0 +1,1161 @@
+        ) -> "Applicative[B]":
+            """
+            Sequences the effects of two Applicative computations,
+            but discards the result of the first.
+            """
+            return cls.apply(cls.const(fa, id), fb)
+
+        @classmethod
+        def keep(
+            cls, fa: "Applicative[A]", fb: "Applicative[B]"
+        ) -> "Applicative[B]":
+            """
+            Sequences the effects of two Applicative computations,
+            but discard the result of the second.
+            """
+            return cls.lift(const, fa, fb)
+
+        @classmethod
+        def revapp(
+            cls, fa: "Applicative[A]", fg: "Applicative[Callable[[A], [B]]]"
+        ) -> "Applicative[B]":
+            """
+            The first computation produces values which are provided
+            as input to the function(s) produced by the second computation.
+            """
+            return cls.lift(lambda a: lambda f: f(a), fa, fg)
+
+        @classmethod
+        def check_identity(cls, fa: "Applicative[A]"):
+            if cls.lift(id, fa) != fa:
+                raise ValueError("Instance violates identity law")
+            return True
+
+        @classmethod
+        def check_homomorphism(cls, a: A, f: Callable[[A], B]):
+            if cls.lift(f, cls.pure(a)) != cls.pure(f(a)):
+                raise ValueError("Instance violates homomorphism law")
+            return True
+
+        @classmethod
+        def check_interchange(cls, a: A, fg: "Applicative[Callable[[A], B]]"):
+            if cls.apply(fg, cls.pure(a)) != cls.lift(lambda g: g(a), fg):
+                raise ValueError("Instance violates interchange law")
+            return True
+
+        @classmethod
+        def check_composition(
+            cls,
+            fg: "Applicative[Callable[[B], C]]",
+            fh: "Applicative[Callable[[A], B]]",
+            fa: "Applicative[A]",
+        ):
+            if cls.apply(fg, cls.apply(fh, fa)) != cls.lift(compose, fg, fh, fa):
+                raise ValueError("Instance violates composition law")
+            return True
+    return (Applicative,)
+
+
+@app.cell(hide_code=True)
+def _(mo):
+    mo.md(
+        r"""
+        # Effectful programming
+
+        Our original motivation for applicatives was the desire to generalise the idea of mapping to functions with multiple arguments. This is a valid interpretation of the concept of applicatives, but from the three instances we have seen it becomes clear that there is also another, more abstract view.
+
+         the arguments are no longer just plain values but may also have effects, such as the possibility of failure, having many ways to succeed, or performing input/output actions. In this manner, applicative functors can also be viewed as abstracting the idea of **applying pure functions to effectful arguments**, with the precise form of effects that are permitted depending on the nature of the underlying functor.
```

Comment:
```suggestion
         The arguments are no longer just plain values but may also have effects, such as the possibility of failure, having many ways to succeed, or performing input/output actions. In this manner, applicative functors can also be viewed as abstracting the idea of **applying pure functions to effectful arguments**, with the precise form of effects that are permitted depending on the nature of the underlying functor.
```
Just a small case typo.